### ISSUE TITLE
[Question] Huge memory consumption

### ISSUE THREAD

**Issue by zean00** [jedisct1/libsodium.js#47]
Hi,

I notice that this library consume quite huge memory portion in a browser (around 30MB in chrome - 3 times my core apps), I've tried to dereference any variable that refer to libsodium but still the memory won't goes down (it's like not being cleaned up by the garbage collector), is it a normal? Is there any way or workaround to free the memory after the the object not used? Will the web worker (https://github.com/jedisct1/libsodium.js/issues/46) style help this issue??

**Comment by jedisct1** [comment]
That's not really expected. What metrics are you looking at?

It seems to consume ~2 Mb (Chrome 49, OSX) here, including the wrapper:

```js
"use strict";
gc();
let heap_before = process.memoryUsage()["heapUsed"],
    sodium = require("libsodium-wrappers.js");
gc();
let heap_after = process.memoryUsage()["heapUsed"];
console.info(`Version: ${sodium.sodium_version_string()}`);
console.info(`Before: ${heap_before} After: ${heap_after} Difference: ${heap_after - heap_before}`);
```
```
Version: 1.0.10
Before: 3344536 After: 5634232 Difference: 2289696
```

**Comment by jedisct1** [comment]
Or maybe the static heap is the culprit here. You can reduce `TOTAL_MEMORY` and see if it helps, but this will prevent password hashing functions from working with the default parameters.

Turning on automatic heap growth could also help, but unfortunately that feature has been removed from asm.js.

**Comment by jedisct1** [comment]
That said, neither the library nor the wrapper attach anything to the global namespace, and when all references are gone, the GC should mark the memory available for new allocations. 
What does the heap profiler show?

**Comment by zean00** [comment]
Hi, Thank you for your fast response.

To be honest I'm new on this memory profiling and confused with all the tooling chrome provided. Maybe I looked at the wrong place. What I did is compare two heap snapshot profile before and after initiate the libsodium object (through require), and I saw  additional object called
backing_store in ArrayBuffer @181339 that using 33 ,554, 496 in size (and it's 78 % of total) on the 2nd snapshot. And this was consistent with the information on a chrome task manager (additional ~30MB in private memory after libsodium initiated), plus this was also the case when I'm deploying it to the android phone (using crosswalk), and check on the running apps manager.

But when I try to check through chrome timeline, it's consistent with the information you gave, the heap growth is only around 2MB and it's goes down once the object is freed. 

I'm not sure which measurement I should follow here, what was that 30MB object? should I worry about that? currently I'm developing a hybrid apps with crosswalk and using this library and I've been trying to reduce memory consumption as much as possible. Thank you for your support :)

**Comment by zean00** [comment]
Based on comparison of heap snapshot statistics before and after libsodium is initialised, there are additional 32,769 KB of Typed Arrays that won't go away even though the variable is already null, and the size is constant overtime, not growing or shrinking. What does it means and isn't that related to memory consumption?

**Comment by zean00** [comment]
Hi,

after playing with the TOTAL_MEMORY variable, I think I could relate that 32MB Typed Array with the configuration of TOTAL_MEMORY in the emscripten compiler. I didn't realise until update the library that 32MB is the old configuration, once my app refer to libsodium 0.3 the Typed Array consume around 50MB (consistent with the latest configuration of emscripten). 

However, after playing with the TOTAL_MEMORY size, here is what I found
- If i set TOTAL_MEMORY below 20MB, the unit test on password hash will fail, so I guess the minimum size is around 20MB
- Even though I set the TOTAL_MEMORY to 20MB, the heap profiler still detect 32MB of Typed Array reside there, is the 32MB is the actual minimum memory?

My question is again, should I worry about this? and is it possible to reduce this size of Typed Arrays ? And how much lower I could reduce it?
